src/components/ClientDialog.tsx

```diff
@@ -0,0 +1,207 @@
+/**
+* ClientDialog.js
+*
+* @2023 Digital Aid Seattle
+*/
+
+import { Button } from "primereact/button";
+import { Dialog } from "primereact/dialog";
+import { InputText } from "primereact/inputtext";
+import { RadioButton } from 'primereact/radiobutton';
+
+import React, { useEffect, useState, useRef} from "react";
+import { TicketType, clientService } from "../services/ClientService";
+import { AnimalType, ClientType, RequestType } from "../types";
+
+const defaultClient: ClientType = {
+  id: null,
+  first_name: '',
+  last_name: '',
+  email: '',
+  phone: '',
+}
+
+const defaultAnimal: AnimalType = {
+  id: null,
+  name: '',
+  species: '',
+  client_id: null,
+}
+
+const defaultRequest: RequestType = {
+  id: null,
+  client_id: null,
+  animal_id: null,
+  service_category: '',
+  source: '',
+  staff_id: null,
+}
+
+const ClientDialog = (props) => {
+
+  const [clientDialog, setClientDialog] = useState(false);
+  const [source, setSource] = useState(TicketType.email);
+  const [request, setRequest] = useState<RequestType>(defaultRequest);
+  const [client, setClient] = useState<ClientType>(defaultClient);
+  const [animal, setAnimal] = useState<AnimalType>(defaultAnimal);
+
+  useEffect(() => {
+    setClientDialog(props.visible)
+  }, [props])
+  
+  const hideClientDialog = () => {
+    props.onClose(undefined);
+    setClient(defaultClient);
+  };
+  
+  const saveClientDialog = () => {
+    // type handle separate to support RadioButton
+
+    setRequest(prevRequest => ({...prevRequest, source: source}));
+    clientService.newRequest(request, client, animal)
+      .then(requestResponse => props.onClose(requestResponse))
+      // TODO - handle all sorts of errors: client exists, animal exists, request exists, etc.
+      .catch(err => props.onClose(null))
+    setRequest(defaultRequest);
+    setClient(defaultClient);
+    setAnimal(defaultAnimal);
+  };
+
+  const timeoutId = useRef(null);
+
+  const autoFillClient = (type: string, value: string) => {
+
+    // REVIEW: This could also be solved with an onBlur call, so that the check
+    // against the DB only happens when user leaves the email field.
+    if (type === 'email') {
+      if (timeoutId.current) {
+        clearTimeout(timeoutId.current);
+      }
+
+      // The request to DB for client will trigger 1sec after staff stops typing
+      timeoutId.current = setTimeout(async () => {
+        try {
+          const clientResponse = await clientService.getClientByEmail(value)
+          console.log(clientResponse)
+          setClient(prevClient => {
+            return {
+              ...prevClient,
+              first_name: clientResponse.first_name || client.first_name,
```

Comment:
Nitpick: Is the empty string a valid value for first_name/last_name/phone? If so, perhaps we could use ?? instead of || in the case where client returns empty string

Reply:
First, last, and phone are required fields so the DB should always have one (if the email already exists). I think a better approach here might be to avoid the setClient call altogether if we dont get a response (changing that now).

Reply:
I agree, with that 👍🏿

But note for the future, the schema doesn't seem to make these required (assuming I was looking in the right place)